kernel/src/table_changes/resolve_dvs.rs

```diff
@@ -0,0 +1,422 @@
+            //   - _Selection_ treemaps (denoted `Treemap_s`) store the indices of selected rows.
+            //     `Treemap_s(0, 1)` means that rows 0 and 1 are _selected_. This is equivalent
+            //     to the selection vector [1, 1, 0] when converted into a boolean vector.
+            // In ordinary scans, only deletion treemaps are used. However in the case of deletion
+            // vector pairs, we generate selection treemaps.
+            //
+            // We use a motivating example to explain the deletion vector resolution. We read
+            // `rm_dv` and `add_dv`, and they are initialized to the empty map by default.
+            // if no deletion vector is given.
+            //  rm_dv  = Treemap_d(0, 1)
+            //  add_dv = Treemap_d(1, 2)
+            //
+            // The result of this commit is:
+            // - row 0 is restored
+            // - row 1 is unchanged (previously deleted)
+            // - row 2 is newly deleted
+            // Thus for this commit we must generate `Treemap_s(0)` for the added rows, and
+            // `Treemap_s(2)` for deleted rows.
+            //
+            //  # Insertion Selection Treemap
+            //  The selection vector of added rows is calculated using set subtraction over deletion
+            //  treemaps `rm_dv - add_dv`. These rows went from set (deleted) in `rm_dv` to unset
+            //  (restored) in the `add_dv`. All other rows are either cancelled in set subtraction,
+            //  or were not selected in either treemap. Hence, they are not selected to
+            //  be in the set of changed rows. Applying this to our deletion treemaps:
+            //  rm_dv - add_dv =
+            //      Treemap_d(0, 1)
+            //    - Treemap_d(1, 2)
+            //    = Treemap_s(0)
+            //  The selection treemap shows that row 0 was inserted
+            //
+            //  # Deletion Selection Treemap
+            //  The selection vector of deleted rows is calculated using `add_dv - rm_dv`. These rows went
+            //  from unset (present) in `rm_dv` to set (deleted) in the `add_dv`. Once again, all
+            //  other rows are either cancelled in set subtraction, or were not set in either treemap.
+            //  Applying this to our deletion vectors:
+            //  add_dv - rm_dv =
+            //      Treemap_d(1, 2)
+            //    - Treemap_d(0, 1)
+            //    = Treemap_s(2)
+            //  The selection treemap shows that row 2 was deleted
+            //
+            //  # Conversion to Selection Vector
+            //  The selection treemap is converted to a selection vector by setting the bits:
+            //      Treemap_s(0) => [true]
+            //      Treemap_s(2) => [false, false, true]
+            //  All other rows are unselected (false).
+            let adds = &rm_dv - &add_dv;
+            let removes = add_dv - rm_dv;
```

Comment:
out of curiosity, do we know the time complexity of these operations?

Reply:
Let n be size of LHS map, m be the size of RHS map.  This is `O(m * log(n) + n)`. This is a set difference, so it iterates over everything in the RHS (O(m)), and for each such element, removes it from the LHS O(log(n)). The adds case has to do a clone of the LHS, so we also add an O(n).

[source](https://docs.rs/roaring/latest/src/roaring/treemap/ops.rs.html#308)